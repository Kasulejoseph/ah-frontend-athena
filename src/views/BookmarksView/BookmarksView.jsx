import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { toast } from "react-toastify";
import { getMyBookmarkedArticlesAction } from "../../actions/getMyBookmarkedArticlesAction";
import BookmarkedList from "../../components/BookmarksList/BookmarksList";

export class BookmarksView extends Component {
  /**
   *
   * @param {} props - props passed into the JSX initialisation of the component
   */
  constructor(props) {
    super(props);
    this.state = { toBeRendered: "No bookmarks available", noData: true };
  }
  componentDidMount() {
    const token = window.localStorage.getItem("token");
    const { getMyBookmarkedArticlesAction } = this.props;
    getMyBookmarkedArticlesAction(token);
  }
  /**
   * Receives props after action is dispatched.
   * @param {} nextProps - properties object that is updated when an
   * action is dispatched and mapStateToProps below is invoked.
   */
  componentWillReceiveProps(nextProps) {
    const bookmarks = nextProps.bookmarks;
    const { history } = this.props;
    if (bookmarks) {
      if (bookmarks.detail) {
        this.setState({ toBeRendered: bookmarks.detail, noData: true });
        toast.error(bookmarks.detail + " Redirecting..."); //session in no longer valid, please login again.
        history.push("/login");
      }
      if (bookmarks.bookmark) {
        this.setState({
          toBeRendered: bookmarks.bookmark,
          noData: false
        });
      }
      if (bookmarks.message) {
        this.setState({ toBeRendered: bookmarks.message, noData: true });
      }
    }
  }
  render() {
    const { noData, toBeRendered } = this.state;
    if (noData) {
      return (
        <h4
          className="d-flex justify-content-center"
          style={{ marginTop: "20%" }}
        >
          You don&apos;t have any bookmarks
        </h4>
      );
    } else {
      return <BookmarkedList articles={toBeRendered} />;
    }
  }
}

BookmarksView.propTypes = {
  history: PropTypes.shape({}),
  bookmarks: PropTypes.shape({}).isRequired,
  getMyBookmarkedArticlesAction: PropTypes.func.isRequired
};
BookmarksView.defaultProps = {
  history: {}
};

/**
 * Filters which state values are mapped to current component's props.
 * @param {} state - state object passed down from the root reducer
 */
export const mapStateToProps = state => ({
  bookmarks: state.bookmarks.bookmarksData
});

export default connect(
  mapStateToProps,
  { getMyBookmarkedArticlesAction }
)(BookmarksView);
