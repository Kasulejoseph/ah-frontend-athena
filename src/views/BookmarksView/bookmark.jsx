import React, { Component } from "react";
import PropTypes from "prop-types";
import { MDBIcon } from "mdbreact";
import { connect } from "react-redux";
import { BookmarkAction } from "../../actions/bookmarkAction/BookmarkAction";
import { UnBookmarkAction } from "../../actions/bookmarkAction/UnbookmarkAction";
import { getMyBookmarkedArticlesAction } from "../../actions/getMyBookmarkedArticlesAction";
import ModalPage from "../../components/Likes/LoginModal";

import "./bookmarks.scss";

const token = window.localStorage.getItem("token");
export class Bookmark extends Component {
  state = {
    is_bookmarked: false,
    bookmark: {},
    colorValue: "blue-grey-text ml-2",
    modal: false
  };

  componentDidMount() {
    const { getMyBookmarkedArticlesAction } = this.props;
    getMyBookmarkedArticlesAction(token);
  }

  componentWillReceiveProps(nextProps) {
    const { data, bookmarkChanges } = nextProps;
    const { getMyBookmarkedArticlesAction } = this.props;
    const { articleId } = this.props;
    if (data.bookmark) {
      this.setState({
        colorValue: "blue-grey-text ml-2",
        is_bookmarked: false
      });
      data.bookmark.map(bookmark => {
        if (bookmark.article === articleId) {
          return this.setState({
            bookmark: bookmark,
            colorValue: "blue-text pr-3",
            is_bookmarked: true
          });
        } else {
          return undefined;
        }
      });
    }
    if (bookmarkChanges) {
      getMyBookmarkedArticlesAction(token);
    }
  }

  toggle = () => {
    const { modal } = this.state;
    this.setState({
      modal: !modal
    });
  };

  onClick = () => {
    const { BookmarkAction, UnBookmarkAction, slug } = this.props;
    const { is_bookmarked, bookmark } = this.state;

    if (window.localStorage.getItem("token") != null) {
      if (is_bookmarked) {
        UnBookmarkAction(bookmark.id);
        getMyBookmarkedArticlesAction(token);
      } else {
        BookmarkAction(bookmark, slug);
        getMyBookmarkedArticlesAction(token);
      }
    } else {
      this.setState({ url: window.location.pathname });
      this.toggle();
    }
  };

  render() {
    const { colorValue, url, modal } = this.state;
    const { history } = this.props;

    return (
      <div className={colorValue}>
        <MDBIcon
          size="2x"
          onClick={this.onClick}
          icon="bookmark"
          className="bookmark-icon"
        />
        <ModalPage
          title="Please login before you can like or dislike the article"
          modal={modal}
          toggle={this.toggle}
          fallback={url}
          history={history}
          md="12"
        />
      </div>
    );
  }
}

Bookmark.propTypes = {
  getMyBookmarkedArticlesAction: PropTypes.func.isRequired,
  BookmarkAction: PropTypes.func.isRequired,
  bookmarkChanges: PropTypes.func.isRequired,
  UnBookmarkAction: PropTypes.func.isRequired,
  articleId: PropTypes.number,
  data: PropTypes.shape({}),
  slug: PropTypes.string,
  history: PropTypes.shape({})
};
Bookmark.defaultProps = {
  articleId: 2,
  data: {},
  slug: "mocked-SLUG",
  history: {}
};
export const mapStateToProps = state => {
  return {
    data: state.bookmarks.bookmarksData,
    bookmarkChanges: state.bookmarks.refresh
  };
};
export default connect(
  mapStateToProps,
  { BookmarkAction, UnBookmarkAction, getMyBookmarkedArticlesAction }
)(Bookmark);
