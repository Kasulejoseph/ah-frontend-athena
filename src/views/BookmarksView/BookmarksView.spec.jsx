import React from "react";
import { shallow } from "enzyme";
import { BookmarksView, mapStateToProps } from "./BookmarksView";

let props = {
  message: "Bookmarks not found",
  getMyBookmarkedArticlesAction: jest.fn(),
  history: { push: jest.fn }
};
const initialState = { bookmarks: { bookmarksData: "mock bookmarks data" } };

describe("Testing the BookmarksView component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<BookmarksView {...props} />);
  });

  it("should render the component properly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  // it("should test that the component exists", () => {
  //   expect(wrapper.exists()).toEqual(true);
  // });

  it("should handle props correctly", () => {
    wrapper.setProps({ ...props });
    expect(wrapper.instance().props.message).toEqual("Bookmarks not found");
  });

  it("should handle the case when bookmarked articles data is not available (no bookmarks)", () => {
    wrapper.setProps({
      bookmarks: { message: "Bookmarks not found" }
    });
    expect(wrapper.state("toBeRendered")).toEqual("Bookmarks not found");
  });

  it("should handle the case when bookmarked articles data is not available (error)", () => {
    wrapper.setProps({
      bookmarks: { detail: "some mocked error" }
    });
    expect(wrapper.state("toBeRendered")).toEqual("some mocked error");
  });

  it("should handle the case when bookmarked articles data is available", () => {
    wrapper.setProps({
      bookmarks: { bookmark: ["mockArticle1", "mockArticle2"] }
    });
    expect(wrapper.state("toBeRendered")).toEqual([
      "mockArticle1",
      "mockArticle2"
    ]);
  });

  it("should map state to props via mapStateToProps", () => {
    expect(mapStateToProps(initialState).bookmarks).toEqual(
      "mock bookmarks data"
    );
  });
});
