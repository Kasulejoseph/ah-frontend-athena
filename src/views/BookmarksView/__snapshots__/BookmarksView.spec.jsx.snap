// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the BookmarksView component should render the component properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookmarksView
    getMyBookmarkedArticlesAction={
      [MockFunction] {
        "calls": Array [
          Array [
            null,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    history={
      Object {
        "push": [Function],
      }
    }
    message="Bookmarks not found"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "You don't have any bookmarks",
      "className": "d-flex justify-content-center",
      "style": Object {
        "marginTop": "20%",
      },
    },
    "ref": null,
    "rendered": "You don't have any bookmarks",
    "type": "h4",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "You don't have any bookmarks",
        "className": "d-flex justify-content-center",
        "style": Object {
          "marginTop": "20%",
        },
      },
      "ref": null,
      "rendered": "You don't have any bookmarks",
      "type": "h4",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
